// @appium/fake-driver/build/lib/commands/alert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const driver_1 = require("appium/driver");
const mixin_1 = require("./mixin");
const AlertMixin = {
    assertNoAlert() {
        if (this.appModel.hasAlert()) {
            throw new driver_1.errors.UnexpectedAlertOpenError();
        }
    },
    assertAlert() {
        if (!this.appModel.hasAlert()) {
            throw new driver_1.errors.NoAlertOpenError();
        }
    },
    /**
     * Get the text of an alert
     */
    async getAlertText() {
        this.assertAlert();
        return this.appModel.alertText();
    },
    /**
     * Set the text of an alert
     */
    async setAlertText(text) {
        this.assertAlert();
        try {
            this.appModel.setAlertText(text);
        }
        catch {
            throw new driver_1.errors.InvalidElementStateError();
        }
    },
    /**
     * Accept an alert
     */
    async postAcceptAlert() {
        this.assertAlert();
        this.appModel.handleAlert();
    },
    /**
     * Dismiss an alert
     */
    async postDismissAlert() {
        return this.postAcceptAlert();
    },
};
(0, mixin_1.mixin)(AlertMixin);
